### PR TITLE
std: Add support for SOCK_SEQPACKET Unix sockets

### PR DESCRIPTION
**Pull request by dkl**
Hi,

this patch adds support for SOCK_SEQPACKET unix sockets to std::os::unix::net. It is based on @rrichardson's [PR for the old unix-socket crate](https://github.com/rust-lang-nursery/unix-socket/pull/25), but ported to the current libstd, with some more unit tests.

`SOCK_SEQPACKET` support was already mentioned in the [RFC for unix-socket](https://github.com/rust-lang/rfcs/blob/master/text/1479-unix-socket.md). The main idea is to add two new structs:
* `UnixSeqpacketListener` similar to `UnixListener` with bind() and accept() for the server-side.
* `UnixSeqpacket` similar to `UnixDatagram` but only with connect(), no bind(), for the client-side. This provides send() and recv() methods, but not send_to()/recv_from() since it's connection-oriented, and also no io::Read/io::Write, due to SOCK_SEQPACKET reads/writes transferring one message at a time, as opposed to being a byte stream.

Without this feature in std, programs have to use plain unsafe libc directly to open the socket, and then possibly use the FromRawFd constructors of UnixDatagram/UnixListener and stick to the  methods that make sense for SOCK_SEQPACKET.

I'm not sure how to handle the stability attributes for this - it's unstable, but can it be assigned to the unix_socket feature, or does it need a new feature name and tracking issue?

### CODE REVIEW

**Comment by pietroalbini** [comment]
Thanks for this PR! The bot failed to assign a reviewer for this, picking someone from the libs team. @sfackler?

File: src/libstd/sys/unix/ext/net.rs [hunk 1]
```diff
@@ -1467,6 +1467,701 @@ impl IntoRawFd for UnixDatagram {
...
+            builder.field("local", &addr);
+        }
+        builder.finish()
+    }
+}
+
+impl UnixSeqpacketListener {
+    /// Creates a new `UnixSeqpacketListener` bound to the specified socket.
+    ///
+    /// Linux provides, as a nonportable extension, a separate "abstract"
```
Comment (sfackler) [inline src/libstd/sys/unix/ext/net.rs]: We don't yet support abstract addresses in the stdlib implementation.

File: src/libstd/sys/unix/ext/net.rs [hunk 2]
```diff
@@ -1467,6 +1467,701 @@ impl IntoRawFd for UnixDatagram {
...
+///             thread::spawn(|| handle_client(socket));
+///         }
+///         Err(err) => {
+///             /* connection failed */
+///             break;
+///         }
+///     }
+/// }
+/// ```
+#[unstable(feature = "unix_socket", issue = "0")]
```
Comment (sfackler) [inline src/libstd/sys/unix/ext/net.rs]: The feature name has to be new for tidy to be happy - `unix_socket_seqpacket` or whatever seems fine.

File: src/libstd/sys/unix/ext/net.rs [hunk 3]
```diff
@@ -1467,6 +1467,701 @@ impl IntoRawFd for UnixDatagram {
...
+            builder.field("peer", &addr);
+        }
+        builder.finish()
+    }
+}
+
+impl UnixSeqpacket {
+    /// Connects to the socket named by `path`.
+    ///
+    /// Linux provides, as a nonportable extension, a separate "abstract"
```
Comment (sfackler) [inline src/libstd/sys/unix/ext/net.rs]: Same as above

**Comment by sfackler** [comment]
LGTM other than the couple of notes above

cc @rust-lang/libs

**Comment by pietroalbini** [comment]
Seems like the documentation contains invalid links on Windows.

```
std\os\unix\net\struct.UnixSeqpacket.html:96: broken link fragment `#method.recv_from` pointing to `std\os\unix\net\struct.UnixSeqpacket.html`
std\os\unix\net\struct.UnixSeqpacket.html:122: broken link fragment `#method.send_to` pointing to `std\os\unix\net\struct.UnixSeqpacket.html`
```

(@TimNN the error bot didn't show up here)

**Comment by dkl** [comment]
Oops, missed those... `send_to/recv_from` methods don't exist for UnixSeqpacket, so the doc comments shouldn't have mentioned them.

**Comment by kennytm** [comment]
[SOCK_SEQPACKET is not available on macOS](https://stackoverflow.com/a/39817558).

**Comment by sfackler** [comment]
This may be blocked until we have the portability lint then, unfortunately. We don't currently have a great way to express "unix other than macOS" right now in the std APIs :(

**Comment by kennytm** [comment]
@sfackler `#[cfg(not(any(target_os = "macos", target_os = "ios"))]`? There is also `std::os::{linux, freebsd, ...}::*`

**Comment by sfackler** [comment]
Sure, we can make it compile, but we don't want methods to be secretly non-portable which a pure-cfg based solution would do.

It doesn't seem reasonable to define identical seqpacket APIs on a separate trait in each of the linux, freebsd, openbsd, android, etc std::os modules.

**Comment by kennytm** [comment]
OK, I'll mark this S-waiting-on-team then.

**Comment by alexcrichton** [comment]
If this is portable everywhere on Unix but osx I'd recommend just putting all this in a crate on crates.io for now (like `unix-socket` started). Once we've got a better story for this sort of portability in libstd we can move it in, but otherwise I don't think it's critical that this live in the standard library.

**Comment by pietroalbini** [comment]
So @rust-lang/libs, is https://github.com/rust-lang/rust/pull/50348#issuecomment-387781928 the decision?

**Comment by alexcrichton** [comment]
Er sorry yes I think I forgot to comment here after triage one week, sorry about that! We discussed this at triage and the conclusion was that we'd prefer if this started out on crates.io and then if there's enough popularity to move it into the standard library we can do so probably without a non-macos module and document that it doesn't work on OSX.

**Pull request closed without merging**